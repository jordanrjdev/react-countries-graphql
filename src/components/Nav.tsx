import { Link } from 'wouter';
import { NavRoute } from '@/models';

const staticRoutes: NavRoute[] = [
  {
    link: 'https://jordanjaramillo.co',
    name: 'Portfolio'
  },
  {
    link: 'https://github.com/jordanrjdev',
    name: 'Github'
  }
];

const Nav = () => {
  const renderNavItem = (route: NavRoute, idRoute: number) => (
    <Link key={idRoute} href={route.link} target="_blank" role="links">
      {route.name}
    </Link>
  );

  return (
    <nav className="grid grid-cols-3 bg-indigo-600 content-center text-white px-10 py-3 space-y-3 md:space-y-0 ">
      <div className="col-span-3 md:col-span-2">
        <Link href="/" className="block text-center md:text-left text-2xl font-bold">
          CFT
        </Link>
      </div>
      <div className="col-span-3 md:col-span-1 flex justify-center font-bold text-xl md:justify-end items-center space-x-3">
        {staticRoutes.map(renderNavItem)}
      </div>
    </nav>
  );
};

export default Nav;
