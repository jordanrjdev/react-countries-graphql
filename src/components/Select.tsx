import { useEffect, useState } from 'react';
import { useCombobox } from 'downshift';

interface SelectType {
  list: string[];
  selectedItems: string[];
  selectItems: (items: string[]) => void;
  onblur: () => void;
  onFocus: () => void;
  placeholder?: string;
  loading: boolean;
  onLoading: JSX.Element;
}

const Select = ({ list, selectedItems, placeholder, selectItems, onblur, onFocus, loading, onLoading }: SelectType) => {
  const [inputItems, setInputItems] = useState(list);
  const { isOpen, getToggleButtonProps, getMenuProps, getInputProps, getComboboxProps, getItemProps } = useCombobox({
    items: inputItems,
    onSelectedItemChange: ({ selectedItem }) => {
      if (!selectedItem) {
        return;
      }
      const index = selectedItems.indexOf(selectedItem);
      if (index > 0) {
        selectItems([...selectedItems.slice(0, index), ...selectedItems.slice(index + 1)]);
      } else if (index === 0) {
        selectItems([...selectedItems.slice(1)]);
      } else {
        selectItems([...selectedItems, selectedItem]);
      }
    },
    selectedItem: null,
    stateReducer: (state: any, actionAndChanges: any) => {
      const { changes, type } = actionAndChanges;
      switch (type) {
        case useCombobox.stateChangeTypes.InputKeyDownEnter:
        case useCombobox.stateChangeTypes.ItemClick:
          return {
            ...changes,
            isOpen: true, // keep menu open after selection.
            highlightedIndex: state.highlightedIndex,
            inputValue: '' // don't add the item string as input value at selection.
          };
        case useCombobox.stateChangeTypes.InputBlur:
          onblur();
          return {
            ...changes,
            inputValue: '' // don't add the item string as input value at selection.
          };
        default:
          return changes;
      }
    },
    onInputValueChange: ({ inputValue }) => {
      if (!inputValue) {
        setInputItems(list);
      } else {
        setInputItems(list.filter((item) => item.toLowerCase().startsWith(inputValue.toLowerCase())));
      }
    }
  });
  const placeholderText = selectedItems?.length
    ? `${selectedItems?.length} ${placeholder?.toLocaleLowerCase() ?? 'elements'} selected`
    : placeholder ?? 'Elements';

  useEffect(() => {
    if (list.length > 0) {
      setInputItems([...list]);
    }
  }, [list]);

  return (
    <div className="relative" onFocus={onFocus} role="select-container">
      <div {...getComboboxProps()} className="flex items-center border border-2 p-2 w-fit rounded-lg my-2" role="select">
        <input placeholder={placeholderText} {...getInputProps()} className="flex-1 focus:outline-none" />
        <button
          role="button"
          type="button"
          {...getToggleButtonProps()}
          className="flex items-center justify-center text-gray-800 focus:outline-none"
        >
          &#8595;
        </button>
      </div>
      <ul
        {...getMenuProps()}
        className={`max-h-80 max-width-300 border-2 border rounded-lg overflow-y-scroll p-0.5 absolute z-10 bg-white w-full${
          !isOpen ? ' hidden' : ''
        }`}
      >
        {isOpen && !loading && inputItems.length < 1 && <p className="text-center text-gray-800">No results found</p>}

        {loading && <div className="w-full h-full flex items-center justify-center">{onLoading}</div>}

        {!loading &&
          inputItems.length > 0 &&
          isOpen &&
          inputItems.map((item, index) => (
            <li
              key={index}
              {...getItemProps({
                item,
                index
              })}
              role="option"
              className="hover:bg-indigo-100 cursor-pointer p-1 flex items-center space-x-1"
            >
              <input type="checkbox" checked={selectedItems.includes(item)} value={item} onChange={() => null} />
              <span />
              {item}
            </li>
          ))}
      </ul>
    </div>
  );
};

export default Select;
