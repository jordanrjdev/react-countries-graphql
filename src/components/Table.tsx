interface TableProps {
  renderHeaders: JSX.Element[];
  renderRows: JSX.Element[];
  loading: boolean;
  onLoading: JSX.Element;
  empty: boolean;
  onEmpty: JSX.Element;
}

export const Table = ({ renderHeaders, renderRows, loading, onLoading, empty, onEmpty }: TableProps) => {
  return (
    <>
      <div className="overflow-auto h-[calc(100vh_-_180px)]">
        <table className="table-auto w-full" role={'table'}>
          <thead className="text-xs font-semibold uppercase text-indigo-400 bg-indigo-100 sticky top-0">
            <tr>{renderHeaders}</tr>
          </thead>
          {!loading && <tbody className="text-sm divide-y divide-indigo-100">{renderRows}</tbody>}
        </table>
        {empty && !loading && <div className="w-full h-full flex justify-center my-10 font-black">{onEmpty}</div>}
        {loading && <div className="w-full h-full flex items-center justify-center">{onLoading}</div>}
      </div>
    </>
  );
};

export const Row = ({ children }: { children: JSX.Element[] }) => {
  return <tr role={'table-row'}>{children}</tr>;
};

export const Header = ({ children }: { children: string }) => {
  return (
    <th role={'table-header'} className="p-2 whitespace-nowrap text-left font-semibold">
      {children}
    </th>
  );
};

export const Cell = ({ children }: { children: string | JSX.Element }) => {
  return <td className="p-2 whitespace-nowrap">{children}</td>;
};

export default {
  Table,
  Row,
  Header,
  Cell
};
