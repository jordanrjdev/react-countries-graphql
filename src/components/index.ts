export { default as Table } from './Table';
export { default as Nav } from './Nav';
export { default as Loading } from './Loading'
export { default as Select } from "./Select"
