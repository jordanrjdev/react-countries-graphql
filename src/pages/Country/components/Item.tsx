import { CountryFragment } from '@/services/graphql';

interface ItemProps {
  country: CountryFragment;
}

const Item = ({ country }: ItemProps) => {
  return (
    <div className="bg-gray-100 rounded-lg p-10 flex flex-col items-center space-y-3">
      <h2 className="text-4xl font-black text-center">
        {country.name} {country.emoji}
      </h2>

      <div className="flex flex-col">
        <p>
          <span className="font-bold">Capital: </span>
          {country.capital ?? '-'}
        </p>

        <p>
          <span className="font-bold">Currency: </span>
          {country.currency ?? '-'}
        </p>

        <p>
          <span className="font-bold">Continent: </span>
          {country.continent.name + ' (' + country.continent.code + ')' ?? '-'}
        </p>

        <p>
          <span className="font-bold">Languages: </span>
          {country.languages.map((language) => language.name).join(', ') ?? '-'}
        </p>
      </div>
    </div>
  );
};

export default Item;
