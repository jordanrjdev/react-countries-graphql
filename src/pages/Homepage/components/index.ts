export { default as Filters } from './Filters'
export { default as Search } from './Search'
export { default as List } from './List'