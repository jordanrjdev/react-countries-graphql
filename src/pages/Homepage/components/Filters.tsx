import { useEffect, useMemo, useState } from 'react';
import { Loading, Select } from '@/components';
import { useContinents, useFilters, useCountries, useCurrencies } from '@/hooks';
import { Continent } from '@/services/graphql';
import { FilterType } from '@/hooks/useFilters';
import Search from './Search';

interface FiltersProps {
  getCountriesByContinent: (continents: string[]) => void;
  getCountriesByName: (name: string) => void;
  getCountriesByCurrencies: (currencies: string[]) => void;
}

let timeOutId: any;

const Filters = () => {
  const { getCountriesByContinent, getAllCountries, applyChangeFilters } = useCountries();
  const { getAllContinents, continents, loading } = useContinents();
  const { getAllCurrencies, currencies } = useCurrencies();
  const { setFilter, search, selectContinent, selectCurrency } = useFilters();

  const renderLoading = useMemo(() => <Loading />, []);

  useEffect(() => {
    applyChangeFilters();
  }, [search]);

  const filterByContinents = () => {
    const codesContinents = continents.filter((continent) => selectContinent.includes(continent.name)).map((continent) => continent.code);
    if (codesContinents.length < 1) {
      getAllCountries();
      return;
    }
    getCountriesByContinent(codesContinents);
  };

  const filterByCurrencies = () => {
    applyChangeFilters();
  };

  const selectContinentChange = (continents: string[]) => {
    setFilter(FilterType.CONTINENT, continents);
  };

  const selectCurrencyChange = (currencies: string[]) => {
    setFilter(FilterType.CURRENCY, currencies);
  };

  const onHandleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (timeOutId) {
      clearTimeout(timeOutId);
    }
    timeOutId = setTimeout(() => {
      setFilter(FilterType.SEARCH, event.target.value);
    }, 500);
  };

  return (
    <div className="flex flex-col md:flex-row items-center md:justify-between">
      <Search placeholder="Search" onChange={onHandleChange} value={search} />

      <div className="flex justfy-start space-x-4">
        <Select
          onFocus={() => {
            if (currencies.length < 1) {
              getAllCurrencies();
            }
          }}
          loading={false}
          onLoading={renderLoading}
          onblur={filterByCurrencies}
          selectItems={selectCurrencyChange}
          selectedItems={selectCurrency}
          list={currencies.map((currency: string) => currency)}
          placeholder="Currencies"
        />

        <Select
          onFocus={() => {
            if (continents.length < 1) {
              getAllContinents();
            }
          }}
          loading={loading}
          onLoading={renderLoading}
          onblur={filterByContinents}
          selectItems={selectContinentChange}
          selectedItems={selectContinent}
          list={continents.map((continent: Continent) => continent.name)}
          placeholder="Continents"
        />
      </div>
    </div>
  );
};

export default Filters;
