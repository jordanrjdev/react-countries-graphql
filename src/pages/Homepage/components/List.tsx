import { Loading } from '@/components';
import { Table } from '@/components';
import { useEffect, useMemo, useState } from 'react';
import { useCountries } from '@/hooks';
import { CountryFragment } from '@/services/graphql';
import { Link } from 'wouter';
import { LinkTo } from '@/styled-components';

const headers: string[] = ['Code', 'Name', 'Continent', 'Currency', 'Actions'];

const List = () => {
  const [loading, setLoading] = useState(false);
  const [list, setList] = useState<CountryFragment[]>([]);
  const { countries, getAllCountries } = useCountries();

  useEffect(() => {
    setLoading(true);
    getAllCountries().then(() => {
      setLoading(false);
    });
  }, []);

  useEffect(() => {
    setList(countries);
  }, [countries]);

  const renderLoading = useMemo(() => <Loading />, []);

  const renderEmpty = useMemo(() => <p className="text-center">No countries found</p>, []);

  const renderHeaders = useMemo(() => headers.map((header) => <Table.Header key={header}>{header}</Table.Header>), [headers]);

  const renderRows = useMemo(
    () =>
      list.map((row: CountryFragment, idRow) => (
        <Table.Row key={idRow}>
          <Table.Cell>{row.code}</Table.Cell>
          <Table.Cell>{row.name}</Table.Cell>
          <Table.Cell>{row.continent.name}</Table.Cell>
          <Table.Cell>{row.currency ?? '-'}</Table.Cell>
          <Table.Cell>
            <Link href={`/country/${row.code}`}>
              <LinkTo>View</LinkTo>
            </Link>
          </Table.Cell>
        </Table.Row>
      )),
    [list]
  );

  return (
    <Table.Table
      loading={loading}
      onLoading={renderLoading}
      renderHeaders={renderHeaders}
      renderRows={renderRows}
      empty={list.length < 1}
      onEmpty={renderEmpty}
    />
  );
};

export default List;
