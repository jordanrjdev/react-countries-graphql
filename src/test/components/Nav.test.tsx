import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { Nav } from '../../components';

describe('<Nav /> Component', () => {
  it('Render nav component', () => {
    render(<Nav />);
    const anchors = screen.getAllByRole('links');
    expect(anchors.length).toBe(2);
    expect(anchors[0]).toHaveTextContent('Portfolio');
    expect(anchors[0]).toHaveAttribute('href', 'https://jordanjaramillo.co');
    expect(anchors[1]).toHaveTextContent('Github');
    expect(anchors[1]).toHaveAttribute('href', 'https://github.com/jordanrjdev');
  });
});

describe('<Nav /> Snapshot', () => {
  it('Should match the snapshot', () => {
    const { asFragment } = render(<Nav />);
    expect(asFragment()).toMatchSnapshot();
  });
});
