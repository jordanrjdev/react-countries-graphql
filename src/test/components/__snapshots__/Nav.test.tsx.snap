// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> Snapshot Should match the snapshot 1`] = `
<DocumentFragment>
  <nav
    class="grid grid-cols-3 bg-indigo-600 content-center text-white px-10 py-3 space-y-3 md:space-y-0 "
  >
    <div
      class="col-span-3 md:col-span-2"
    >
      <a
        class="block text-center md:text-left text-2xl font-bold"
        href="/"
      >
        CFT
      </a>
    </div>
    <div
      class="col-span-3 md:col-span-1 flex justify-center font-bold text-xl md:justify-end items-center space-x-3"
    >
      <a
        href="https://jordanjaramillo.co"
        role="links"
        target="_blank"
      >
        Portfolio
      </a>
      <a
        href="https://github.com/jordanrjdev"
        role="links"
        target="_blank"
      >
        Github
      </a>
    </div>
  </nav>
</DocumentFragment>
`;
